block-layer/codecs/dag-json.md

```diff
@@ -42,13 +51,49 @@ The Bytes kind is represented as an object with `"bytes"` as key and a [Multibas
 
 ```javascript
 {"/": String /* Base58 encoded CIDv0 or Multibase Base32 encoded CIDv1 */}
 ```
+
+### The Reserved Namespace
+
+Maps with the first key of `"/"` are considered the **reserved namespace** in DAG-JSON as they are used to represent Bytes and Links. There are special rules that restrict certain data forms from being properly encoded in DAG-JSON. These rules allow for the clean representation of Bytes and Links as well as efficient operation of tokenizing decoders. A tokenizing decoder should not need to buffer and back-track more than 4 tokens upon detection of a map that is not properly encoding Links or Bytes.
+
+Specifically, data with the following forms are strictly not valid DAG-JSON and should be rejected by encoders and decoders:
+
+ * Maps with more than one key, where the first key is `"/"` and its value is a string. e.g. `{"/":"foo","bar":"baz"}`.
+   - Where a key exists that sorts before `"/"`, the map is valid, e.g. `{"0bar":"baz","/":"foo"}`.
+   - Where the value of the `"/"` entry is not a string, the map is valid, e.g. `{"/":true,"bar":"baz"}`
+ * Maps where the first key is `"/"` and its value is a map with more than one key where the first key of the inner map is `"bytes"` whose value is a string. e.g. `{"/":{"bytes":"foo","bar":"baz"}}`.
+   * Where a key exists in the inner map that sorts before `"bytes"`, the map is valid, e.g. `{"/":{"abar":"baz","bytes":"foo"}}`.
+   * Where the value of the inner map's `"bytes"` entry is not a string, the map is valid, e.g. `{"/":{"bytes":true},"bar":"baz"}`.
+ * Maps with more than one key, where the first key is `"/"` and its value is a map where the first key of the inner map is `"bytes"` whose value is a string. e.g. `{"/":{"bytes":"foo"},"bar":"baz"}`.
+   - Where a key exists that sorts before `"/"`, the map is valid, e.g. `{"0bar":"baz","/":{"bytes":"foo"}}`.
+   - Where the value of the `"bytes"` entry in the inner map is not a string, the map is valid, e.g. `{"/":{"bytes":true},"bar":"baz"}`
+
+Further:
+
+ * A map with the single key `"/"`, whose value is a string, must contain a valid CIDv0 base58 string or CIDv1 base32 string. Such a map whose string does not properly represent such a CID should should be rejected as invalid DAG-CBOR.
+ * A map with the single key `"/"`, whose value is a map with the single key `"bytes"`, whose value is a string, must contain a valid multibase base64 encoded byte array. Such a construction whose string does not properly represent a multibase base64 encoded byte array should be rejected as invalid DAG-CBOR.
+
+There is no mechanism for escaping otherwise valid JSON data that takes these forms. For this reason, it is recommended that the `"/"` key should be avoided in Data Model maps where DAG-JSON may be used in order to avoid such conflicts.
+
+## Implementations
```

Comment:
This section isn't quite correct but we could get it all cleaned up before merging so that it is correct. `@ipld/dag-json` currently has the same caveats as `ipld-dag-json` as well as a couple of bugs. go-ipld-prime doesn't do bytes at all yet I think, maybe not even links, last I checked. It got stalled on the reserved space discussions. There was a branch with the beginnings of the work necessary but I can't see that anymore.